src/app/shared/helpers/data-pager.ts

```diff
@@ -0,0 +1,81 @@
+import { map, Observable, ReplaySubject, shareReplay, switchMap } from 'rxjs';
+import { fetchingState, FetchState, readyState } from './state';
+import { mapToFetchState } from '../operators/state';
+
+export function canLoadMorePagedData<T>(list: T[], limit: number): boolean {
+  /**
+   * If list length is same as limit, it means we can still fetch.
+   * For the edge case when the list length is same as limit but there's no more items, it will be
+   * solved by itself at next call because the list will then be empty.
+   */
+  return list.length === limit;
+}
+
+interface PagerOptions<T> {
+  fetch: (lastElement?: T) => Observable<T[]>,
+  batchSize: number,
+  /**
+   * If an error occurs while loading more elements, we do not want the list to appear as broken.
+   * The state will remain ready with old data and this callback will be triggered, allowing to display feedback
+   */
+  onLoadMoreError: (error: Error) => void,
+}
+
+export class DataPager<T> {
+  private trigger$ = new ReplaySubject<void>(1);
+
+  canLoadMore = true;
+  state$: Observable<FetchState<T[]>> = this.trigger$.pipe(
+    switchMap(() => this.options.fetch(this.lastElement).pipe(
+      mapToFetchState(),
+      map(state => {
+        // Case 1: First fetch -> return state as is, and when ready populate our accumulator in prevision of further fetches
+        if (this.acc.length === 0) {
```

Comment:
using `this` in a operator block 😬

Reply:
> For the pager, IMO that's not valid Rx as you modify the instance freely from the operator.

I agree, I initially used "pure" rxjs and it was complicated to understand in my opinion. Because of that I considered it was the same case as custom operators in which you can have side-effects because they are encapsulated in the operator.
Now I'll check if the the bug you've found it related to make sure it is not the problem.

Reply:
> I considered it was the same case as custom operators in which you can have side-effects because they are encapsulated in the operator.
 
I understand, but IMO it is not the same, here you change the instance of an object, not just the scope of a single limited operator. But as I wrote as conclusion: I have to think about it :-)